compiler/rustc_codegen_llvm/src/debuginfo/metadata.rs

```diff
@@ -762,47 +636,54 @@ pub fn type_metadata<'ll, 'tcx>(cx: &CodegenCx<'ll, 'tcx>, t: Ty<'tcx>) -> &'ll
-            // Also make sure that we already have a `TypeMap` entry for the unique type ID.
-            let Some(metadata_for_uid) = type_map.find_metadata_for_unique_id(unique_type_id) else {
-                bug!(
-                    "expected type metadata for unique \
-                            type ID '{}' to already be in \
-                            the `debuginfo::TypeMap` but it \
-                            was not. (Ty = {})",
-                    type_map.get_unique_type_id_as_string(unique_type_id),
-                    t
-                );
-            };
-
-            match type_map.find_metadata_for_type(t) {
-                Some(metadata) => {
-                    if metadata != metadata_for_uid {
+            // Make sure that we really do have a `TypeMap` entry for the unique type ID.
+            let metadata_for_uid =
+                match debug_context(cx).type_map.find_metadata_for_unique_id(unique_type_id) {
+                    Some(metadata) => metadata,
+                    None => {
                         bug!(
-                            "mismatch between `Ty` and \
-                                   `UniqueTypeId` maps in \
-                                   `debuginfo::TypeMap`. \
-                                   UniqueTypeId={}, Ty={}",
-                            type_map.get_unique_type_id_as_string(unique_type_id),
+                            "expected type metadata for unique \
+                               type ID '{:?}' to already be in \
+                               the `debuginfo::TypeMap` but it \
+                               was not. (Ty = {})",
+                            unique_type_id,
                             t
                         );
                     }
-                }
-                None => {
-                    type_map.register_type_with_metadata(t, metadata);
-                }
-            }
+                };
+
+            debug_assert_eq!(metadata_for_uid as *const _, metadata as *const _);
         } else {
-            type_map.register_type_with_metadata(t, metadata);
-            type_map.register_unique_id_with_metadata(unique_type_id, metadata);
+            debug_context(cx).type_map.register_unique_id_with_metadata(unique_type_id, metadata);
         }
     }
 
     metadata
 }
 
+fn recursion_marker_type<'ll, 'tcx>(cx: &CodegenCx<'ll, 'tcx>) -> &'ll DIType {
+    *debug_context(cx).recursion_marker_type.get_or_init(move || {
+        unsafe {
+            // The choice of type here is pretty arbitrary -
+            // anything reading the debuginfo for a recursive
+            // type is going to see *something* weird - the only
+            // question is what exactly it will see.
+            //
+            // FIXME: the name `<recur_type>` does not fit the naming scheme
+            //        of other types.
+            let name = "<recur_type>";
+            llvm::LLVMRustDIBuilderCreateBasicType(
```

Comment:
Pre-existing but I wonder if we should create this as a `void*` so at least a debugger will show the address?

Reply:
Should I address this in this PR or separately?

Reply:
Maybe just leave a comment if you agree? I don't think we need to fix it right now.

Reply:
Good idea!